src/FindStmtCost.cpp

```diff
@@ -0,0 +1,930 @@
+                           << "FindStmtCost::get_computation_cost: " << print_node(node)
+                           << "node not found in stmt_cost map"
+                           << "\n\n";
+            return 0;
+        }
+    } else {
+        if (inclusive) cost = it->second.computation_cost_inclusive;
+        else
+            cost = it->second.computation_cost_exclusive;
+    }
+
+    if (cost < 0) {
+        internal_error << "\n"
+                       << "FindStmtCost::get_computation_cost: " << print_node(node)
+                       << "computation_cost_exclusive not set (cost is: " << cost << ")"
+                       << "\n\n";
+    }
+
+    return cost;
+}
+int FindStmtCost::get_data_movement_cost(const IRNode *node, bool inclusive) const {
+    if (node == nullptr) {
+        internal_error << "\n"
+                       << "FindStmtCost::get_data_movement_cost: node is nullptr"
+                       << "\n\n";
+    }
+
+    auto it = stmt_cost.find(node);
+    IRNodeType type = node->node_type;
+    int cost = -1;
+
+    if (it == stmt_cost.end()) {
+        // sometimes, these constant values are created on the whim in
+        // StmtToViz.cpp - set cost_node to be fresh StmtCost to avoid crashing
+        if (type == IRNodeType::IntImm || type == IRNodeType::UIntImm ||
+            type == IRNodeType::FloatImm || type == IRNodeType::StringImm) {
+            cost = NORMAL_NODE_DMC;
+        }
+
+        // this happens when visualizing cost of else_case in StmtToViz.cpp
+        else if (type == IRNodeType::Variable) {
+            const Variable *var = (const Variable *)node;
+            if (var->name == "canIgnoreVariableName") {
+                cost = NORMAL_NODE_DMC;
+            }
+        }
+
+        // else, error
+        else {
+            internal_error << "\n"
+                           << "FindStmtCost::get_data_movement_cost: " << print_node(node)
+                           << "node not found in stmt_cost map"
+                           << "\n\n";
+            return 0;
+        }
+    } else {
+        if (inclusive) cost = it->second.data_movement_cost_inclusive;
+        else
+            cost = it->second.data_movement_cost_exclusive;
+    }
+
+    if (cost < 0) {
+        internal_error << "\n"
+                       << "FindStmtCost::get_data_movement_cost: " << print_node(node)
+                       << "data_movement_cost_exclusive not set (cost is: " << cost << ")"
+                       << "\n\n";
+    }
+
+    return cost;
+}
+
+int FindStmtCost::get_if_node_cost(const IRNode *op, bool inclusive, bool is_computation) const {
+    if (op->node_type != IRNodeType::IfThenElse) {
+        internal_error << "\n"
+                       << "FindStmtCost::get_if_node_cost: " << print_node(op)
+                       << "node is not IfThenElse"
+                       << "\n\n";
+    }
+
+    int cost;
+    const IfThenElse *if_then_else = dynamic_cast<const IfThenElse *>(op);
```

Comment:
Don't use dynamic_cast<> in Halide code -- we aren't always build with RTTI enabled, so this won't work -- and there is a *much* better way to do this with Halide IR nodes, namely, the `.as<>()` method (which is heavily optimized):

```
    const IfThenElse *if_then_else = op->as<IfThenElse>();  // nullptr if op is not of this type
```

Reply:
It's been a while, but it looks like `as<T>` is no longer an interface of `IRNode`; I see one in `IRHandle`, though.

Reply:
Is it just a matter of locally wrapping he `IRNode*` into an `IRHandle` just to have access to the `as<T>()` interface?

Reply:
get_if_node_cost should take a const IfThenElse *, and the cast should happen in the caller as a C-style cast. If you wrap it in an IRHandle to use `as` it'll do a pointless refcount.

Reply:
How about something like `string GetReadWrite::print_node(const IRNode *node) const { ... }`?